python/revng/internal/cli/_commands/schema_diff.py

```diff
@@ -0,0 +1,86 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import yaml
+from revng.internal.cli.commands_registry import Command, CommandsRegistry
+import copy
+
+
+class SchemaDiffCommand(Command):
+    def __init__(self):
+        super().__init__(
+            ("model", "schema-diff"), "List the changes between an old and a new schema"
+        )
+
+    def register_arguments(self, parser):
+        parser.add_argument("old_schema", metavar="OLD_SCHEMA")
+        parser.add_argument("new_schema", metavar="NEW_SCHEMA")
+
+    def _restructured_schema(self, schema):
+        ret = {}
+
+        for defn in schema:
+            name = defn["name"]
+
+            subdefs_list = defn["members"] if defn["type"] == "enum" else defn["fields"]
+
+            subdefs = {}
+            for sd in subdefs_list:
+                sdname = sd["name"]
+                sdprops = {k: v for k, v in sd.items() if k != "name"}
+
+                subdefs[sdname] = sdprops
+
+            props = {k: v for k, v in defn.items() if k not in {"name", "type", "fields", "members"}}
+
+            ret[name] = defn["type"], props, subdefs
+
+        return ret
+
+    def run(self, options):
+        args = options.parsed_args
+
+        with open(args.old_schema) as file:
+            old_schema = yaml.safe_load(file)
```

Comment:
What prevents us from using `tuple_tree_generator.Schema` and have strong typing?
The idea of parsing the model as unstructured data was limited to the model, we don't need to do it for its schema.

Reply:
From what I see so far: nothing is preventing us, I just subconsciously went with the same approach as with the model. I will rewrite it using the `Schema` class.

Reply:
Since `tuple_tree_generator.Schema` is under `scripts`, not `python/revng`, we cannot use it without some refactoring first.

@mrjackv once [hinted](https://mattermost.rev.ng/revng/pl/orhknydh8fydtezwp7qij9z5hw) mentioned that we must have `revng model schema-diff` under `python/revng`, so the only way to proceed from here would be to move `Schema` and related utilities from `script` to `python/revng`. However, this seems like quite a drastic change to make.

What's your opinion? Should we proceed with this change?

Reply:
The stuff in `scripts/tuple_tree_generator/tuple_tree_generator/schema` can be quite easily moved under `revng/internal`, I don't see a problem with it.
`PYTHONPATH` will need to be "conditionally" adjusted in `share/revng/cmake/TupleTreeGenerator.cmake` (maybe it should be moved somewhere else?) since the generation of model stuff happens before the revng-python stuff is installed.

Reply:
> maybe it should be moved somewhere else?

Can you please elaborate on this? What needs to moved?

Reply:
The `share` directory is installed as part of the `revng` installation, this means that potentially other users could include `TupleTreeGenerator.cmake` in their project. Due to this there's two ways to proceed:
1. We keep supporting `TupleTreeGenerator.cmake` being included by other projects, and need to fix `PYTHONPATH` but only if we're doing things within revng (I'm not sure if there's a straightforward way to do this in cmake)
2. We decide that `TupleTreeGenerator.cmake` should not be installed anymore, move it somewhere else and patch `PYTHONPATH` all the time

What do you think @aleclearmind ?

Reply:
With the current resolution we will not be using the `Schema` wrapper here.

I will close this and move the conversation to the sibling [here](https://github.com/revng/revng/pull/438#discussion_r1975669768).